### PR TITLE
remove innerHTML from text layer

### PR DESCRIPTION
**Pull request by fjakobs**
In old browsers (especially IE6/7) `innerHTML` used to be much faster than the DOM. Therefore the Ace text layer is almost exclusively based on `innerHTML`. In modern browser this has changed dramatically and now the DOM is much faster. This PR changes the DOM layer to no longer use `innerHTML`. In addition to the performance benefit this also removes a potential source for XSS issues.

### CODE REVIEW

**Comment by lennartcl** [comment]
How does this relate to https://github.com/ajaxorg/ace/pull/3372?

**Comment by fjakobs** [comment]
@lennartcl I wasn't aware of the branch. Thanks for pointing this out. It it essentially the same except that @nightwing 's  PR also does it for the marker layer.

Both are not ready because tests are still failing.

File: lib/ace/virtual_renderer.js [hunk 1]
```diff
@@ -838,8 +838,10 @@ var VirtualRenderer = function(container, theme) {
             this.$updateScrollBarV();
             if (changes & this.CHANGE_H_SCROLL)
                 this.$updateScrollBarH();
-            this.$gutterLayer.element.style.marginTop = (-config.offset) + "px";
-            this.content.style.marginTop = (-config.offset) + "px";
+            
+            dom.translate(this.$gutterLayer.element, 0, -config.offset);
```
Comment (nightwing) [inline lib/ace/virtual_renderer.js]: On windows transforms break antialiasing, and on my mac laptop screen i don't see significant difference in performance, compared to adding `contain:strict` on layer elements.
Is different more significant on larger screens?

Comment (fjakobs) [reply lib/ace/virtual_renderer.js]: In theory transform will bypass paint and go straight into compositing. Can you try https://stackoverflow.com/questions/6492027/css-transform-jagged-edges-in-chrome and see if that fixes it?

File: lib/ace/virtual_renderer.js [hunk 2]
```diff
@@ -428,18 +427,21 @@ var VirtualRenderer = function(container, theme) {
...
-        if (this.session.getUseWrapMode() && this.adjustWrapLimit()) {
-            this.$loop.schedule(this.CHANGE_FULL);
-        } else if (this.$size.$dirty) {
-            this.$loop.schedule(this.CHANGE_FULL);
-        } else {
-            this.$computeLayerConfig();
-            this.$loop.schedule(this.CHANGE_MARKER);
-        }
+        var that = this;
+        setTimeout(function() {
```
Comment (nightwing) [inline lib/ace/virtual_renderer.js]: Why is this needed, if it is to not call gutter.offsetWidth, i think we can safely remove it https://github.com/ajaxorg/ace/commit/e701c27e30ddf47365dcaee9f67c3a2d23f2bb92#diff-1d3adf2538ef07e28972236f3a26311aR431

File: lib/ace/virtual_renderer.js (hunk 1 above)
Comment (nightwing) [reply lib/ace/virtual_renderer.js]: It indeed bypasses painting, and that breaks subpixel antialiasing. Setting `backface-visibility:hidden;` or using other hacks does not seem to help.

Setting `will-change: transform` or `contain: strict` changes behavior of marginTop/top to skip paint as well, but sadly it has similar effect on antialiasing.

On windows the positive effect of this is very small anyway, because we do not have smooth scrolling, and when scroll amount is large enough to change lines we trigger full layout and repaint by removing the line

File: lib/ace/layer/gutter.js [hunk 3]
```diff
@@ -228,6 +191,154 @@ var Gutter = function(parentEl) {
...
+                className += " ace_open";
+            if (foldWidget.className != className)
+                foldWidget.className = className;
+
+            var height = config.lineHeight + "px";
+            if (foldWidget.style.height != height)
+                foldWidget.style.height = height;
+                
+            if (foldWidget.style.display != "block")
+                foldWidget.style.display = "block";
```
Comment (nightwing) [inline lib/ace/layer/gutter.js]: this needs to be `inline-block`, to not go to the next line, so may be better to use `.display =  ""` instead

File: lib/ace/virtual_renderer.js (hunk 2 above)
Comment (fjakobs) [reply lib/ace/virtual_renderer.js]: It was needed because it would render the change in the same frame causing the frame to take too long to render. With this change the update is applied in the the following frame.

File: lib/ace/layer/gutter.js (hunk 3 above)
Comment (fjakobs) [reply lib/ace/layer/gutter.js]: changed

File: lib/ace/virtual_renderer.js (hunk 2 above)
Comment (nightwing) [reply lib/ace/virtual_renderer.js]: This works well while scrolling, and another render is going to happen anyway, but when pressing enter it render same thing twice, and there is small visible jump.

File: lib/ace/layer/marker.js [hunk 4]
```diff
@@ -90,14 +101,13 @@ var Marker = function(parentEl) {
...
-                } else {
-                    this.drawSingleLineMarker(html, range, marker.clazz + " ace_start" + " ace_br15", config);
-                }
+                this.drawSingleLineMarker(html, range, marker.clazz + " ace_start" + " ace_br15", config);
             }
         }
-        this.element.innerHTML = html.join("");
+        if (this.i !=-1) {
+            while (this.i < this.element.childElementCount)
+                this.element.lastChild.remove();
```
Comment (nightwing) [inline lib/ace/layer/marker.js]: remove is not defined on ie11, should we add it to our polifill similar to https://github.com/c9/core/blob/master/plugins/c9.ide.ui/lib_apf.js#L1502-L1506?
